[BoundsSafety] Improve the soft trap call interface to reduce codesize impact

delcypher opened the pull request:
This patch makes several changes motivated to reduce the code size impact at call sites to runtime functions.

* `__bounds_safety_soft_trap_c` function no longer takes an integer argument and has been renamed to remove the `_c` suffix. The corresponding `call-with-code` option to `-fbounds-safety-soft-traps=` has been renamed to `call-minimal`. We can reduce codeisze by avoiding writing constants into the `x0` on arm64 which saves us 1 instruction (4 bytes) per call. The original motivation for passing an integer code was to aid debugging in the scenario where debug info is missing. This scenario isn't the primary use case and codesize is far more important so removing the argument seems like the right call.

* The runtime functions now use the `preserve_all` calling convention on x86_64 and arm64. This calling convention means that some registers that are normally caller-saved become callee-saved. This means that these registers do not need to be spilled at callsites which should reduce some code size overhead. This isn't the perfect calling convention because the argument passing registers could also be made callee saved currently aren't but such a calling convention isn't currently implemented in LLVM so this is the best we can do for now. A hidden flag (`-bounds-safety-soft-trap-preserve-all-cc=false`) to turn off this behavior is included. It is primarly there to make it easy to measure the codesize impact without having to rebuild the compiler.

Note: While this is an ABI breaking change
`__CLANG_BOUNDS_SAFETY_SOFT_TRAP_API_VERSION` has not been bumped because there are currently aren't any adopters of the soft trap mode yet.

Another change this patch makes is to explicitly add the `llvm::Attribute::NoMerge` attribute on calls to the soft trap runtime. I haven't come across a concrete case where this is necessary but given that we want this behavior it seemed like a bug that we weren't explicitly requesting it. While this should probably be in its own commit I put it in the same commit to reduce the number of time I needed to regenerating the test cases because some require manual patching.

rdar://164341162

delcypher committed:
[BoundsSafety] Improve the soft trap call interface to reduce codesize impact

This patch makes several changes motivated to reduce the code size
impact at call sites to runtime functions.

* `__bounds_safety_soft_trap_c` function no longer takes an integer argument and
  has been renamed to remove the `_c` suffix. The corresponding
  `call-with-code` option to `-fbounds-safety-soft-traps=` has been
  renamed to `call-minimal`. We can reduce codeisze by avoiding writing
  constants into the `x0` on arm64 which saves us 1 instruction (4
  bytes) per call. The original motivation for passing an integer code
  was to aid debugging in the scenario where debug info is missing.
  This scenario isn't the primary use case and codesize is far more
  important so removing the argument seems like the better design
  choice.

* The runtime functions now use the `preserve_all` calling convention
  on x86_64 and arm64. This calling convention means that some registers
  that are normally caller-saved become callee-saved. This means that
  these registers do not need to be spilled at callsites which should
  reduce some code size overhead. This isn't the perfect calling
  convention because the argument passing registers could also be made
  callee saved currently aren't but such a calling convention isn't
  currently implemented in LLVM so this is the best we can do for now.
  A hidden flag (`-bounds-safety-soft-trap-preserve-all-cc=false`) to
  turn off this behavior is included. It is primarly there to make it
  easy to measure the codesize impact without having to rebuild the
  compiler.

Note: While this is an ABI breaking change
`__CLANG_BOUNDS_SAFETY_SOFT_TRAP_API_VERSION` has not been bumped
because there are currently aren't any adopters of the soft trap mode
yet.

rdar://164341162

hnrklssn commented:
> We can reduce codeisze by avoiding writing constants into the x0 on arm64 which saves us 1 instruction (4 bytes) per call. The original motivation for passing an integer code was to aid debugging in the scenario where debug info is missing. This scenario isn't the primary use case and codesize is far more important so removing the argument seems like the right call.

Does this not also allow for call sites to be merged for even greater code size savings?

> Another change this patch makes is to explicitly add the llvm::Attribute::NoMerge attribute on calls to the soft trap runtime. I haven't come across a concrete case where this is necessary but given that we want this behavior it seemed like a bug that we weren't explicitly requesting it.

Do we actually always want this?

delcypher commented:
>> We can reduce codeisze by avoiding writing constants into the x0 on arm64 which saves us 1 instruction (4 bytes) per call. The original motivation for passing an integer code was to aid debugging in the scenario where debug info is missing. This scenario isn't the primary use case and codesize is far more important so removing the argument seems like the right call.

> Does this not also allow for call sites to be merged for even greater code size savings?

In theory yes, but in practice no. So far I've never seen the compiler try to merge the calls and I suspect this probably because the control flow resuming (i.e. carry on as if the bounds check had passed) after the call makes merging the calls more difficult. If the compiler did merge the calls it would have to keep track of where it needs to jump back to after the call and doing a transformation like that doesn't sound like it would be profitable for performance or code size. Also, we don't actually want the calls merged because this would make things really hard to debug. The whole point of soft-trap mode is to make traps non-fatal but also give the developer enough data to debug the problem so it can be fixed.

>> Another change this patch makes is to explicitly add the llvm::Attribute::NoMerge attribute on calls to the soft trap runtime. I haven't come across a concrete case where this is necessary but given that we want this behavior it seemed like a bug that we weren't explicitly requesting it.

> Do we actually always want this?

I think so but I think it's mistake to do it in the same commit. I was just being lazy. I'll split it out.

hnrklssn commented:
Oh right, the whole reason the traps are straightforward to merge is because they terminate the program. D'oh!

delcypher commented on clang/lib/CodeGen/CGExpr.cpp:
```diff
@@ -4838,35 +4846,51 @@ void CodeGenFunction::EmitTrapCheck(llvm::Value *Checked,
+      }
+
+      TrapCall = EmitNounwindRuntimeCall(TrapFunc, RuntimeCallArgs);
+      TrapCall->addFnAttr(llvm::Attribute::NoMerge);
```
@fhahn Should we be putting any the `memory(...)` attributes on the soft trap runtime function?

Implementations of the soft trap runtime are supposed to log the event (in whatever way they see fit) and return. While a soft trap runtime may maintain its own global state that is not supposed to be visible to the rest of the program.

I ask because I wonder if code would be better optimized (e.g. avoid spilling recently modified values in the caller  that are currently in registers back to memory just in case the called function would read it) if there were attributes communicating how the soft trap runtime function uses memory.

delcypher commented:
@swift-ci test llvm

hnrklssn reviewed (approved):
LGTM

delcypher commented:
Turns out we shouldn't land this yet. It seems the `preserve_all` attribute is having no effect on the finally generate assembly. I'm investigating now.

delcypher committed:
[BoundsSafety] Mark calls to soft trap runtime functions as NoMerge
In practice I haven't seen the optimizer merge calls to runtime trap
functions. This is likely because the control flow continuing after the
call makes optimization more difficult compared to hard traps where
control flow does not resume.

However, given that soft trap runtime requires that the traps be
debuggable we should explicitly tell the optimizer to not merge the
calls just in-case the optimizer gets smarter in the future.

rdar://164341162

delcypher commented:
Ok I figured it out. The original pass only set `preserve_all` on the function the IR but didn't set it on the call site. It needs to be set on the call site to get different codegen at the callsite. I've fixed that and updated the tests to account for this.
